src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,273 @@
+                _byteBuffer.Array,
+                _byteBuffer.Offset,
+                _byteBuffer.Array,
+                0,
+                _byteBuffer.Count);
+            int readBytes =
+#if NETSTANDARD2_0
+                await _stream.ReadAsync(_byteBuffer.Array, _byteBuffer.Count, _byteBuffer.Array.Length, cancellationToken).ConfigureAwait(false);
+#else
+                await _stream.ReadAsync(_byteBuffer.Array.AsMemory(_byteBuffer.Count), cancellationToken).ConfigureAwait(false);
+#endif
+            _byteBuffer = new ArraySegment<byte>(_byteBuffer.Array, 0, _byteBuffer.Count + readBytes);
+
+            Debug.Assert(_charBuffer.Count == 0, "We should only expect to read more input chars once all buffered content is read");
+
+#if NETSTANDARD2_0
+            _decoder.Convert(_byteBuffer.Array, _byteBuffer.Offset, _byteBuffer.Count, _charBuffer.Array, 0, _charBuffer.Array.Length,
+                flush: readBytes == 0, out int bytesUsed, out int charsUsed, out _);
+#else
+            _decoder.Convert(
+                _byteBuffer.AsSpan(),
+                _charBuffer.Array,
+                flush: readBytes == 0,
+                out int bytesUsed,
+                out int charsUsed,
+                out _);
+#endif
+
+            Debug.Assert(_charBuffer.Array != null);
+
+            _byteBuffer = _byteBuffer.Slice(bytesUsed);
+            _charBuffer = new ArraySegment<char>(_charBuffer.Array, 0, charsUsed);
+        }
+
+        private static void ThrowArgumentOutOfRangeException(byte[] buffer, int offset, int count)
+        {
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (offset < 0 || offset >= buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                _disposed = true;
+
+                Debug.Assert(_charBuffer.Array != null);
+                ArrayPool<char>.Shared.Return(_charBuffer.Array);
+
+                Debug.Assert(_byteBuffer.Array != null);
+                ArrayPool<byte>.Shared.Return(_byteBuffer.Array);
+
+                Debug.Assert(_overflowBuffer.Array != null);
+                ArrayPool<byte>.Shared.Return(_overflowBuffer.Array);
```

Comment:
This stream is being potentially handed out to user code.  As with my comment the write stream should null these out, the read stream must as well; otherwise, user code using this stream after it's been disposed is going to lead to using a buffer that's already been returned to the pool.

cc: @GrabYourPitchforks

Reply:
Can you elaborate, how is it possible to hand out `TranscodingReadStream` to user code? I will null out the arrays, though.

Reply:
My comment is from before I understood that you were only handing it to the JsonSerializer.